src/psyclone/psyir/nodes/container.py

```diff
@@ -171,21 +171,85 @@ def node_str(self, colour=True):
         return f"Container[{self.name}]\n"
 
-    def get_routine_psyir(self, name):
-        '''Returns the PSyIR for the routine with the given name, or None
-        if a routine with this name does not exist.
+    def get_routine_psyir(self, name, allow_private=False,
+                          check_wildcard_imports=False):
+        '''
+        Searches the Container for a definition of the named routine.
+
+        NOTE: if the named routine corresponds to a generic interface then this
+        method will return None. You will need to use `resolve_routine` first
+        to find the names of the routines that the interface resolves to.
 
-        :param str name: name of the routine to find.
+        If it is not found and the routine is named in an import statement
+        then the search is continued in the named Container. Failing that,
+        all wildcard imports are checked.
 
-        :returns: the PSyIR Routine instance of the subroutine, or None if
-            there is no routine with that name in this container.
-        :rtype: Union[None, psyclone.psyir.nodes.Routine]
+        :param str name: the name of the Routine for which to search.
+        :param bool allow_private: whether the Routine is permitted to have
+            a visibility of PRIVATE.
+
+        :returns: the PSyIR of the named Routine if found, otherwise None.
+        :rtype: :py:class:`psyclone.psyir.nodes.Routine` | NoneType
 
         '''
-        name = name.lower()
-        for routine in self.walk(Routine):
-            if routine.name.lower() == name:
-                return routine
+        rname = name.lower()
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.routine import Routine
+        from psyclone.psyir.symbols.symbol import Symbol, SymbolError
+        # pylint: enable=import-outside-toplevel
+        for node in self.children:
+            if isinstance(node, Routine) and node.name.lower() == rname:
+                # Check this routine is public
+                routine_sym = self.symbol_table.lookup(node.name)
+                if (allow_private or
+                        routine_sym.visibility == Symbol.Visibility.PUBLIC):
+                    return node
+                # The Container does not contain the expected Routine or the
+                # Routine is not public.
+
+        # Look in the import that names the routine if there is one.
+        table = self.symbol_table
+        try:
+            routine_sym = table.lookup(rname)
+        except KeyError:
+            # Routine does not exist in the SymbolTable.
+            routine_sym = None
+
+        if routine_sym and isinstance(routine_sym, GenericInterfaceSymbol):
+            # The routine is actually an interface to one or more routines.
+            # This should be handled outside this routine (e.g. by using
+            # `resolve_routine` first).
+            return None
+
+        if routine_sym and routine_sym.is_import:
```

Comment:
The `if (allow_private or routine_sym.visibility == Symbol.Visibility.PUBLIC)` also needs to be done here and with the wilcard imports below because we can have:
```
module my_mod
   use modA, only: sub1
   use modB  ! contains sub2
   private :: sub1, sub2
end module
```
We want get_routine_psyir() by default to not find sub1 and sub2

Reply:
Good point. I've added the check after we lookup the RoutineSymbol. Further down, before we look at wildcard imports, I've added a check on the default visibility of the table as, without an explicit Symbol entry, that will determine whether an imported symbol is visible. I'll need to add tests for this too.